### PR TITLE
GitHub Actions: Run Clippy on tests

### PR DESCRIPTION
**Pull request by petertseng**
Supports https://github.com/exercism/rust/issues/659

Note that if we merge this PR, it implies we accept the burden of keeping ourselves in line with Clippy so long as we don't revert it. So we should make sure we want to accept that burden before merging.

Should discuss: Is it possible that an upstream Clippy update will cause previously-succeeding CI to fail on the same commit because of new Clippy lints that have been added? If so, what, if anything, do we want to do about that?

A PR that breaks compilation is included because the behavior of bin/test-exercise changed and it's important to check that it still propagates its exit code.

This will also fail until being rebased on https://github.com/exercism/rust/pull/1003, but that is just as well - it verifies that the Clippy check works correctly.

### CODE REVIEW

File: .github/workflows/tests.yml
Comment (coriolinus) [inline .github/workflows/tests.yml:102]: Right now this is under the `compilation` key, which runs it on all steps in the matrix. I don't think that's necessary. 

Probably the best solution is adding it to its own build job, kind of like the benchmarking on nightly. We might in the future decide that we want to run it on its own matrix, so we can get warnings when new lints are introduced on the beta channel before they hit stable.

**Comment by petertseng** [comment]
> This will also fail until being rebased on #1003, but that is just as well - it verifies that the Clippy check works correctly.

Nope, it doesn't, because of our only running on changed exercises. I'll just push new commits then. But we should still not merge this until #1003 is merged.

File: _test/check-exercises.sh
Comment (coriolinus) [inline _test/check-exercises.sh:16]:
```suggestion
if [ -z "$DENYWARNINGS" ] || [ -z "$CLIPPY" ]; then
```

Equivalent, but I like the more explicit form better.

**Comment by petertseng** [comment]
> Should discuss: Is it possible that an upstream Clippy update will cause previously-succeeding CI to fail on the same commit because of new Clippy lints that have been added? If so, what, if anything, do we want to do about that?

My assessment is: That is possible, just like how previously-succeeding CI might fail when a new Rust toolchain release gets promoted to beta. We are not doing anything special to deal with that situation either, just fixing them as they come, so I suppose we'd do the same here and not do anything special.

**Review by coriolinus — COMMENTED** [review]
> Is it possible that an upstream Clippy update will cause previously-succeeding CI to fail on the same commit because of new Clippy lints that have been added?

Yes, and in fact is fairly common. For example, the lint warning about redundant `'static` in `const A_STR: &'static str` was added shortly after it became legal to omit the explicit lifetime from that sort of declaration. The best thing I can think of to mitigate this is to tailor precisely what jobs are run:

- when GHA runs on a PR, it should only run clippy on files modified by that PR
- when GHA runs on `master` branch, it should run clippy on all files

In both cases, we probably want to run it both on `stable` and `beta` rustc so that we get some forewarning when a new lint is going to become a problem. We may wish to configure it that it's not a failing error if clippy reports something on `beta`, but if the build doesn't actually break, I'm not sure we'd actually notice if `beta` started reporting new issues.

File: _test/check-exercises.sh
Comment (petertseng) [reply _test/check-exercises.sh]: Sure, I will be more explicit. `-z "$DENYWARNINGS$CLIPPY"` is only true if both are empty, so I think it should be && instead of ||.

**Comment by petertseng** [comment]
> > Is it possible that an upstream Clippy update will cause previously-succeeding CI to fail on the same commit because of new Clippy lints that have been added?
> 
> The best thing I can think of to mitigate this is to tailor precisely what jobs are run:
>
> * when GHA runs on a PR, it should only run clippy on files modified by that PR
> * when GHA runs on master branch, it should run clippy on all files

All right, that seems like a good idea. The "violate clippy by adding return" and subsequent revert give me assurance that the first bullet point has been achieved, and I believe that since it's going through _test/check-exercises.sh that the second bullet point will also be achieved, but 100% certainty on that will only be gained after merging.

**Comment by petertseng** [comment]
If we want either or both of these two new checks "Clippy (stable)" or "Clippy (beta)" to be required, we will need to remember to ask an admin to make them required after this is merged.

**Comment by petertseng** [comment]
... though I see that no checks are currently marked as required, so maybe we don't need to do that and the current situation works just fine for us

File: .github/workflows/tests.yml
Comment (petertseng) [reply .github/workflows/tests.yml:102]: OK. I just went ahead and added the matrix. stable and beta.

Comment (petertseng) [reply .github/workflows/tests.yml:102]: ... which makes me realise and I never actually removed this Install clippy item. and the fact that Clippy still succeeded despite that means that it's being installed by default anyway.

**Review by coriolinus — APPROVED** [review]

**Comment by petertseng** [comment]
Oh, here's an idea. Merging this before #1003 would give us a rare opportunity to test the "running on master runs on every exercise" point. I'd still want to be able to merge #1003 right after that gets confirmed, so I'll still wait for approval on #1003.

Squash of this PR is planned.

**Comment by petertseng** [comment]
> running on master runs on every exercise

Nope, didn't happen, because of TRAVIS_PULL_REQUEST. That has convinced me it's time to expunge that. PR coming very soon.

File: bin/test-exercise
Comment (petertseng) [inline bin/test-exercise:89]: Notice! I have discovered that this has proven error-prone. I will be submitting a PR to correct that in a few hours.

**Pull request merged**
